pkg/rbd/controllerserver.go

```diff
@@ -462,66 +508,128 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
+			return nil, status.Errorf(codes.NotFound, "source Volume ID %s not found", parentVolID)
+		}
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	// Check if source volume was created with required image features for snaps
+	if !hasSnapshotFeature(rbdVol.ImageFeatures) {
+		return nil, status.Errorf(codes.InvalidArgument, "volume(%s) has not snapshot feature(layering)", parentVolID)
+	}
+	return rbdVol, nil
 }
 
-func (cs *ControllerServer) doSnapshot(rbdSnap *rbdSnapshot, cr *util.Credentials) (err error) {
-	err = createSnapshot(rbdSnap, cr)
-	// If snap creation fails, even due to snapname already used, fail, next attempt will get a new
-	// uuid for use as the snap name
+func (cs *ControllerServer) createSnapFromClone(rbdVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials) error {
+
+	var (
+		err     error
+		snapErr error
+		volErr  error
+	)
+	found, err := checkSnapExists(rbdSnap, cr)
 	if err != nil {
-		klog.Errorf("failed to create snapshot: %v", err)
-		return status.Error(codes.Internal, err.Error())
+		if _, ok := err.(util.ErrSnapNameConflict); ok {
+			return status.Error(codes.AlreadyExists, err.Error())
+		}
+		return status.Errorf(codes.Internal, err.Error())
 	}
-	defer func() {
+
+	if found {
+		klog.Infof("found temp snapshot image %s ", rbdSnap.RequestName)
+		// check clone volume Exist
+		_, err = checkVolExists(rbdVol, cr)
 		if err != nil {
-			errDefer := deleteSnapshot(rbdSnap, cr)
-			if errDefer != nil {
-				klog.Errorf("failed to delete snapshot: %v", errDefer)
-				err = fmt.Errorf("snapshot created but failed to delete snapshot due to"+
-					" other failures: %v", err)
-			}
-			err = status.Error(codes.Internal, err.Error())
+			return status.Error(codes.Internal, err.Error())
 		}
-	}()
-
-	err = protectSnapshot(rbdSnap, cr)
+		klog.Infof("found temp clone image %s for snapshot %s", rbdVol.RequestName, rbdSnap.RequestName)
+		goto deleteSnap
+	}
+	err = reserveSnap(rbdSnap, cr)
 	if err != nil {
-		klog.Errorf("failed to protect snapshot: %v", err)
 		return status.Error(codes.Internal, err.Error())
 	}
-	defer func() {
-		if err != nil {
-			errDefer := unprotectSnapshot(rbdSnap, cr)
-			if errDefer != nil {
-				klog.Errorf("failed to unprotect snapshot: %v", errDefer)
-				err = fmt.Errorf("snapshot created but failed to unprotect snapshot due to"+
-					" other failures: %v", err)
-			}
-			err = status.Error(codes.Internal, err.Error())
-		}
-	}()
 
-	err = getSnapshotMetadata(rbdSnap, cr)
+	// createsnapshot
+	snapErr = createSnapshot(rbdSnap, cr)
+
+	if snapErr != nil {
+		errDefer := undoSnapReservation(rbdSnap, cr)
+		if errDefer != nil {
+			klog.Warningf("failed undoing reservation of snapshot: %s (%s)", rbdSnap.RequestName, errDefer)
+		}
+		return snapErr
+	}
+	err = reserveVol(rbdVol, cr)
 	if err != nil {
-		klog.Errorf("failed to fetch snapshot metadata: %v", err)
 		return status.Error(codes.Internal, err.Error())
 	}
 
+	volErr = restoreSnapshot(rbdVol, rbdSnap, cr)
```

Comment:
Nit: can we rename `restoreSnapshot` to clarify that it's actually just performing an "rbd clone"? i.e. `cloneSnapshot`?